optics-core/src/Optics/Plated.hs

```diff
@@ -0,0 +1,352 @@
+-- compositionality. Given two single transformations @f@ and @g@, you can
+-- construct @\\a -> f a '<|>' g a@ which performs both rewrites until a fixed
+-- point.
+rewriteOf :: Is k A_Setter => Optic k is a b a b -> (b -> Maybe a) -> a -> b
+rewriteOf l f = go
+  where
+    go = transformOf l (\x -> maybe x go (f x))
+{-# INLINE rewriteOf #-}
+
+-- | Rewrite recursively over part of a larger structure.
+rewriteOn :: (Is k A_Setter, Plated a) => Optic k is s t a a -> (a -> Maybe a) -> s -> t
+rewriteOn b = over b . rewrite
+{-# INLINE rewriteOn #-}
+
+-- | Rewrite recursively over part of a larger structure using a specified 'Setter'.
+rewriteOnOf
+  :: (Is k1 A_Setter, Is k2 A_Setter)
+  => Optic k1 is1 s t a b
+  -> Optic k2 is2 a b a b
+  -> (b -> Maybe a) -> s -> t
+rewriteOnOf b l = over b . rewriteOf l
+{-# INLINE rewriteOnOf #-}
+
+-- | Rewrite by applying a monadic rule everywhere you can. Ensures that the
+-- rule cannot be applied anywhere in the result.
+rewriteM :: (Plated a, Monad m) => (a -> m (Maybe a)) -> a -> m a
+rewriteM = rewriteMOf plate
+{-# INLINE rewriteM #-}
+
+-- | Rewrite by applying a monadic rule everywhere you recursing with a
+-- user-specified 'Traversal'.
+--
+-- Ensures that the rule cannot be applied anywhere in the result.
+rewriteMOf
+  :: (Is k A_Traversal, Monad m)
+  => Optic k is a b a b
+  -> (b -> m (Maybe a)) -> a -> m b
+rewriteMOf l f = go
+  where
+    go = transformMOf l (\x -> f x >>= maybe (return x) go)
+{-# INLINE rewriteMOf #-}
+
+-- | Rewrite by applying a monadic rule everywhere inside of a structure located by a user-specified 'Traversal'.
+-- Ensures that the rule cannot be applied anywhere in the result.
+rewriteMOn
+  :: (Is k A_Traversal, Plated a, Monad m)
+  => Optic k is s t a a
+  -> (a -> m (Maybe a)) -> s -> m t
+rewriteMOn b = traverseOf b . rewriteM
+{-# INLINE rewriteMOn #-}
+
+-- | Rewrite by applying a monadic rule everywhere inside of a structure located
+-- by a user-specified 'Traversal', using a user-specified 'Traversal' for
+-- recursion. Ensures that the rule cannot be applied anywhere in the result.
+rewriteMOnOf
+  :: (Is k1 A_Traversal, Is k2 A_Traversal, Monad m)
+  => Optic k1 is1 s t a b
+  -> Optic k2 is2 a b a b
+  -> (b -> m (Maybe a)) -> s -> m t
+rewriteMOnOf b l = traverseOf b . rewriteMOf l
+{-# INLINE rewriteMOnOf #-}
+
+-------------------------------------------------------------------------------
+-- Universe
+-------------------------------------------------------------------------------
+
+-- | Retrieve all of the transitive descendants of a 'Plated' container,
+-- including itself.
+universe :: Plated a => a -> [a]
+universe = universeOf plate
+{-# INLINE universe #-}
+
+-- | Given a 'Fold' that knows how to locate immediate children, retrieve all of
+-- the transitive descendants of a node, including itself.
+universeOf :: Is k A_Fold => Optic' k is a a -> a -> [a]
+universeOf l = go
+  where
+    go a = a : foldMapOf l go a
```

Comment:
I know `lens` doesn't do this either, but is there a reason not to define `universe*` fucntions in terms of `cosmos*` and `toListOf`?

Reply:
I have no idea, sadly I never used these functions so have no intuition behind them.